matrix-e2e.c

```diff
@@ -287,6 +287,88 @@ static MatrixOlmSession *find_olm_session(MatrixConnectionData *conn,
     return result;
 }
 
+static void free_matrix_olm_session(MatrixOlmSession *msession,
+                                    gboolean free_session)
+{
+    g_free(msession->sender_id);
+    g_free(msession->sender_key);
+    if (free_session) {
+        g_free(msession->session);
+    }
+}
+
+/* Save a new OLM session into the database */
+static MatrixOlmSession *store_olm_session(MatrixConnectionData *conn,
+                                           OlmSession *session,
+                                           const char *sender,
+                                           const char *sender_key)
+{
+    MatrixOlmSession *cur_entry = g_new(MatrixOlmSession, 1);
+    size_t pickle_len = olm_pickle_session_length(session);
+    gchar *pickle = g_malloc(pickle_len+1);
+    sqlite3_stmt *dbstmt = NULL;
+
+    pickle_len = olm_pickle_session(session, "!", 1, pickle, pickle_len);
+    if (pickle_len == olm_error()) {
+        purple_debug_warning("matrixprpl",
+                             "%s: Failed to pickle session for %s/%s: %s\n",
+                             __func__, sender, sender_key,
+                             olm_session_last_error(session));
+        goto err;
+    }
+    pickle[pickle_len] = '\0';
+
+    cur_entry->sender_id = g_strdup(sender);
+    cur_entry->sender_key = g_strdup(sender_key);
+    cur_entry->session = session;
+    cur_entry->unique = g_get_monotonic_time();
```

Comment:
I think you can get away without this; sqlite has a rowid property. after doing the insert, you can call `sqlite3_last_insert_rowid` which gives you an int64; then you can do things like `select rowid from olmsessions` and `update olmsessions where rowid=?`.